Check for SSH_ERROR on channel_read_nonblocking

Qalthos opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
`libssh.ssh_channel_read_nonblocking()` can also return `SSH_ERROR`, which is helpfully defined as -1. This is not a valid buffer length.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

Qalthos committed:
Check for SSH_ERROR on channel_read_nonblocking

Qalthos committed:
Get error string and add to exception

Qalthos committed:
Add changelog

webknjaz reviewed (commented):
Any chance to reproduce this in a test?

Qalthos commented:
I'd love to, but I'm not sure what's causing it yet. I'm not even entirely sure this is the issue, although we have seen read_nonblocking return a negative number and this seems like the most likely cause

webknjaz commented:
@Qalthos do you have a traceback to post? It'd be useful to see more details.

webknjaz reviewed (dismissed):
@Qalthos the CI failure is legit: https://github.com/ansible/pylibssh/pull/168/checks?check_run_id=1856047223#step:10:1050 — Cython is unable to compile your change.

Qalthos commented:
Ah, dang, yeah, wrong session object

This is the traceback this is coming from:

```
Traceback (most recent call last):
  File "/tmp/ansible-test-2zxwj1dw/ansible/utils/jsonrpc.py", line 46, in handle_request
    result = rpc_method(*args, **kwargs)
  File "/home/zuul/.ansible/collections/ansible_collections/arista/eos/plugins/cliconf/eos.py", line 174, in get
    check_all=check_all,
  File "/tmp/ansible-test-2zxwj1dw/ansible/plugins/cliconf/__init__.py", line 127, in send_command
    resp = self._connection.send(**kwargs)
  File "/home/zuul/.ansible/collections/ansible_collections/ansible/netcommon/plugins/connection/network_cli.py", line 315, in wrapped
    return func(self, *args, **kwargs)
  File "/home/zuul/.ansible/collections/ansible_collections/ansible/netcommon/plugins/connection/network_cli.py", line 905, in send
    command, prompt, answer, newline, prompt_retry_check, check_all
  File "/home/zuul/.ansible/collections/ansible_collections/ansible/netcommon/plugins/connection/network_cli.py", line 861, in receive
    check_all,
  File "/home/zuul/.ansible/collections/ansible_collections/ansible/netcommon/plugins/connection/network_cli.py", line 767, in receive_libssh
    data = self._ssh_shell.read_bulk_response()
  File "src/pylibsshext/channel.pyx", line 124, in pylibsshext.channel.Channel.read_bulk_response
  File "src/pylibsshext/channel.pyx", line 126, in pylibsshext.channel.Channel.read_bulk_response
  File "src/pylibsshext/channel.pyx", line 105, in pylibsshext.channel.Channel.read_nonblocking
SystemError: Negative size passed to PyBytes_FromStringAndSize
```
https://cce9f6cf9a2ca4962c1a-5f251b929e97de4291a98ef762a28228.ssl.cf1.rackcdn.com/168/bfbc0bb970904e0adae03bd0ba9afa120e2e347b/check/ansible-test-network-integration-eos-network_cli-libssh-python36-stable210-scenario02/45b6669/controller/ansible-debug.html#l73102

webknjaz commented:
@Qalthos please rebase, I've fixed the CI in #170.

Qalthos commented:
For posterity, the traceback with this patch applied is now:

```
Traceback (most recent call last):
  File "[...]/ansible/lib/ansible/utils/jsonrpc.py", line 46, in handle_request
    result = rpc_method(*args, **kwargs)
  File "[...]/ansible_collections/arista/eos/plugins/cliconf/eos.py", line 168, in get
    return self.send_command(
  File "[...]/ansible/lib/ansible/plugins/cliconf/__init__.py", line 127, in send_command
    resp = self._connection.send(**kwargs)
  File "[...]/ansible_collections/ansible/netcommon/plugins/connection/network_cli.py", line 315, in wrapped
    return func(self, *args, **kwargs)
  File "[...]/ansible_collections/ansible/netcommon/plugins/connection/network_cli.py", line 904, in send
    response = self.receive(
  File "[...]/ansible_collections/ansible/netcommon/plugins/connection/network_cli.py", line 855, in receive
    response = self.receive_libssh(
  File "[...]/ansible_collections/ansible/netcommon/plugins/connection/network_cli.py", line 767, in receive_libssh
    data = self._ssh_shell.read_bulk_response()
  File "src/pylibsshext/channel.pyx", line 128, in pylibsshext.channel.Channel.read_bulk_response
  File "src/pylibsshext/channel.pyx", line 130, in pylibsshext.channel.Channel.read_bulk_response
  File "src/pylibsshext/channel.pyx", line 108, in pylibsshext.channel.Channel.read_nonblocking
pylibsshext.errors.LibsshChannelReadFailure: Socket error: disconnected
```

Which is the expected result

webknjaz commented:
@Qalthos have you considered a dedicated error for this? I was thinking along the lines of inheriting `ConnectionError` and similar.

Qalthos committed:
Add custom exception for channel read errors

It's a Python object and not a C one, because I can't figure out how to
subclass ConnectionError from C

Qalthos commented:
Well, dang. It can't be `ConnectionError`, then. Any other ideas that are 2.7-compatible?

webknjaz commented:
@Qalthos `ConnectionError` is a subclass of `OSError`. We could conditionally inherit that. Or maybe follow this example https://github.com/ansible/pylibssh/commit/726a58384f41383d81cdc890cf4b9cf06dd7abab.

P.S. ansible-core is to drop Python 2 on controller (https://github.com/ansible/ansible/issues/72668) and also we've been talking that this lib is going to be packaged with Python 3 under RHEL8 — should we talk about dropping py2?

webknjaz commented on src/pylibsshext/errors.pyx:
```diff
@@ -33,11 +33,18 @@ cdef class LibsshException(Exception):
     pass
 
+
+class LibsshChannelReadFailure(LibsshChannelException, ConnectionError):
```
@Qalthos this is roughly what you're looking for:

```suggestion
try:  # py3
    connection_error_exc = ConnectionError
except NameError:  # py2
    connection_error_exc = OSError


class LibsshChannelReadFailure(LibsshChannelException, connection_error_exc):
```

webknjaz commented on src/pylibsshext/errors.pyx:
```diff
@@ -33,11 +33,18 @@ cdef class LibsshException(Exception):
 
+
+class LibsshChannelReadFailure(LibsshChannelException, ConnectionError):
+    pass
```
Could you please add a docstring instead of `pass`?

Qalthos commented:
> P.S. ansible-core is to drop Python 2 on controller ([ansible/ansible#72668](https://github.com/ansible/ansible/issues/72668)) and also we've been talking that this lib is going to be packaged with Python 3 under RHEL8 — should we talk about dropping py2?

If we can get away with it (and I think we should be able to given that paramiko is still supported) I'd much rather drop Python 2 than continue to try to work around it.

Qalthos committed:
Replace pass with docstring

webknjaz commented:
> If we can get away with it (and I think we should be able to given that paramiko is still supported) I'd much rather drop Python 2 than continue to try to work around it.

Sure, but this is beyond the scope of this PR and will probably take time to confirm. Maybe somebody still wants it with py27 on PyPI...

Qalthos committed:
Update LibsshChannelReadFailure for Python 2 support

Co-authored-by: Sviatoslav Sydorenko <wk.cvs.github@sydorenko.org.ua>

webknjaz reviewed (approved):
LGTM but it's probably a good idea to get @ganeshrn's approval.

ganeshrn reviewed (approved):
lgtm.
Thank you!
